core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1637,6 +1640,83 @@ public void sync(Translog.Location location) {
+        final List<Tuple<Translog.Location, Consumer<Exception>>> candidates = new ArrayList<>();
+        if (promised) {
+            // if we are responsible we can't wait for others to make space - we have to sync until we can add to the queue
+            while (fsyncQueue.offer(new Tuple<>(location, syncListener)) == false) {
+                drainAndSync(candidates);
+            }
+        } else {
+            // in this case we are not responsible and can just block until there is space
+            try {
+                fsyncQueue.put(new Tuple<>(location, syncListener));
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+
+        // here we have to try to make the promise again otherwise there is a race when a thread puts an entry without making the promise
+        // while we are draining that mean we might exit below too early in the while loop if the drainAndSync call is fast.
+        if (promised || promiseSync.tryAcquire()) {
+            // since we made the promise to sync we gotta do it here at least once
+            drainAndSync(candidates);
+            promiseSync.release(); // now to ensure we are passing it on we release the promise so another thread can take over
+            while (fsyncQueue.isEmpty() == false && promiseSync.availablePermits() > 0) {
+                // yet if the queue is not empty AND nobody else has yet made the promise to take over we continue syncing to disk
+                drainAndSync(candidates);
+            }
+        }
+    }
+
+    private void drainAndSync(List<Tuple<Translog.Location, Consumer<Exception>>> candidates) {
+        candidates.clear();
+        fsyncQueue.drainTo(candidates);
+        syncList(candidates);
+    }
+
+    private void syncList(List<Tuple<Translog.Location, Consumer<Exception>>> candidates) {
+        Exception exception = null;
+        if (candidates.isEmpty() == false) {
+            try {
+                final Engine engine = getEngine();
+                engine.getTranslog().ensureSynced(candidates.stream().map(Tuple::v1));
+            } catch (EngineClosedException ex) {
+                // that's fine since we already synced everything on engine close - this also is conform with the methods
+                // documentation
+            } catch (IOException ex) { // if this fails we are in deep shit - fail the request
+                logger.debug("failed to sync translog", ex);
+                // this exception is passed to all listeners - we don't retry. if this doesn't work we are in deep shit
+                exception = ex;
+            }
+        }
+        for (Tuple<Translog.Location, Consumer<Exception>> tuple : candidates) {
+            Consumer<Exception> consumer = tuple.v2();
+            try {
+                consumer.accept(exception);
+            } catch (Exception ex) {
+                logger.warn("translog sync consumer failed", ex);
```

Comment:
I'm not sure giving the exception to the calling thread is all that useful though. It doesn't have anything to do with the other consumer. Beyond switching this to `logger.error` I'm not sure what else we can do.